File: src/main/java/at/rueckgr/spotlight/service/ImageDownloaderService.java
package at.rueckgr.spotlight.service;

import at.rueckgr.spotlight.model.DownloadableImage;
import at.rueckgr.spotlight.util.SpotlightException;
import org.apache.commons.io.IOUtils;
import org.apache.http.HttpEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;

import java.io.IOException;

public class ImageDownloaderService {
    public byte[] fetchImage(final DownloadableImage downloadableImage) {
        final CloseableHttpClient httpClient = HttpClients.custom().disableDefaultUserAgent().build();

        final HttpGet httpGet = new HttpGet(downloadableImage.getUrl());
        try (final CloseableHttpResponse response = httpClient.execute(httpGet)) {
            final int statusCode = response.getStatusLine().getStatusCode();
            if (statusCode != 200) {
                throw new SpotlightException(String.format("Fail to retrive image content: %s", statusCode));
            }
            HttpEntity entity = response.getEntity();
            return IOUtils.toByteArray(entity.getContent());
        }
        catch (IOException e) {
            throw new SpotlightException(e);
        }
    }
}

File: src/main/kotlin/at/rueckgr/spotlight/service/ImageDownloaderService.kt
package at.rueckgr.spotlight.service

import at.rueckgr.spotlight.model.DownloadableImage
import at.rueckgr.spotlight.util.Logging
import at.rueckgr.spotlight.util.SpotlightException
import at.rueckgr.spotlight.util.logger
import org.apache.commons.io.IOUtils
import org.apache.hc.client5.http.classic.methods.HttpGet
import org.apache.hc.client5.http.impl.classic.HttpClients

class ImageDownloaderService : Logging {
    fun fetchImage(downloadableImage: DownloadableImage): ByteArray {
        val httpClient = HttpClients.custom().disableDefaultUserAgent().build()

        logger().debug("Fetching image ${downloadableImage.url}")

        httpClient.execute(HttpGet(downloadableImage.url)).use {
            val statusCode = it.code
            if (statusCode != 200) {
                throw SpotlightException(String.format("Fail to retrieve image content: %s", statusCode))
            }
            return IOUtils.toByteArray(it.entity.content)
        }
    }
}
